### PR TITLE
Hoist

### PR DESCRIPTION
**Pull request by quardbreak**
Ports hoist from Aurora station. Useful thing for cargo or places which have open space.

### CODE REVIEW

File: code/modules/multiz/hoist.dm [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+	qdel(src)
+
+/obj/effect/hoist_hook
+	name = "hoist clamp"
+	desc = "A clamp used to lift people or things."
+	icon = 'icons/obj/hoists.dmi'
+	icon_state = "hoist_hook"
+	can_buckle = TRUE
+	anchored = TRUE
+
```
Comment (MistakeNot4892) [inline code/modules/multiz/hoist.dm:33]: Probably want `simulated = FALSE` here.

File: code/modules/multiz/hoist.dm [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+	set category = "Object"
+	set src in range(1)
+
+	if (!ishuman(usr))
+		return
+
+	if (isobserver(usr) || usr.incapacitated())
+		return
+	if (!usr.check_dexterity(DEXTERITY_GRIP, silent = TRUE)) // thanks nanacode
+		to_chat(usr, SPAN_NOTICE("You stare cluelessly at \the [src]."))
```
Comment (MistakeNot4892) [inline code/modules/multiz/hoist.dm]: This message isn't very informative about the actual reason the check is failing.

Comment (MistakeNot4892) [reply code/modules/multiz/hoist.dm]: Also, SPAN_WARNING for warning messages.

File: code/game/objects/structures/catwalk.dm [hunk 3]
```diff
@@ -128,6 +128,12 @@
 							break
 					update_icon()
 
+/obj/structure/catwalk/hoist_act(turf/dest)
+	for(var/A in loc)
+		var/atom/movable/AM = A
+		AM.forceMove(dest)
```
Comment (comma) [inline code/game/objects/structures/catwalk.dm]: probably needs anchored / simulated checks so you're not lifting lighting overlays

File: code/modules/multiz/hoist.dm [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	name = "hoist kit"
+	desc = "A setup kit for a hoist that can be used to lift things. The hoist will deploy in the direction you're facing."
+	icon = 'icons/obj/hoists.dmi'
+	icon_state = "hoist_case"
+
+/obj/item/hoist_kit/attack_self(mob/user)
+	if (!do_after(usr, (2 SECONDS), src))
+		return
+
+	new /obj/structure/hoist(get_turf(user), user.dir)
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: needs to transfer fingerpints when going between forms

File: code/modules/multiz/hoist.dm [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+	user.visible_message(
+		SPAN_WARNING("[user] deploys the hoist kit!"),
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: why is it a warning, doesn't sound so dangerous / failure-like

File: code/modules/multiz/hoist.dm [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+/obj/effect/hoist_hook/attack_hand(mob/living/user)
+	return // no, bad
+
+/obj/effect/hoist_hook/MouseDrop_T(atom/movable/AM,mob/user)
+	if (issilicon(user))
+		return
+
+	if (!AM.simulated || AM.anchored)
+		to_chat(user, SPAN_NOTICE("You can't do that."))
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: now this should be warning, and maybe adding bit more inf oabout what 'that' is like name of object

File: code/modules/multiz/hoist.dm [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+		return
+	if (source_hoist.hoistee)
+		to_chat(user, SPAN_NOTICE("\The [source_hoist.hoistee] is already attached to \the [src]!"))
+		return
+	source_hoist.attach_hoistee(AM)
+	user.visible_message(
+		SPAN_DANGER("[user] attaches \the [AM] to \the [src]."),
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: danger should be reserved for direct combat danger, this is notice probably, warning at most

File: code/modules/multiz/hoist.dm [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+		SPAN_DANGER("You attach \the [AM] to \the [src]."),
+		SPAN_DANGER("You hear something clamp into place."))
+
+/obj/structure/hoist/proc/attach_hoistee(atom/movable/AM)
+	if (get_turf(AM) != get_turf(source_hook))
+		AM.forceMove(get_turf(source_hook))
+	hoistee = AM
```
Comment (comma) [inline code/modules/multiz/hoist.dm:56]: need to subscrube to events of hoistee's deletion

File: code/modules/multiz/hoist.dm [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+		return
+	if (!dest.Adjacent(source_hoist.hoistee))
+		return
+
+	source_hoist.check_consistency()
+
+	var/turf/desturf = dest
+	source_hoist.hoistee.forceMove(desturf)
+	usr.visible_message(
+		SPAN_DANGER("[usr] detaches \the [source_hoist.hoistee] from the hoist clamp."),
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: again overreacting with the span

File: code/modules/multiz/hoist.dm [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+		SPAN_DANGER("You hear something unclamp."))
+	source_hoist.release_hoistee()
+
+// This will handle mobs unbuckling themselves.
+/obj/effect/hoist_hook/unbuckle_mob()
+	. = ..()
+	if (. && !QDELETED(source_hoist))
+		var/mob/M = .
+		source_hoist.hoistee = null
+		fall(M)	// fuck you, you fall now!
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: that's not how fall is called, should be `M.fall(get_turf(src))`

File: code/modules/multiz/hoist.dm [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+		release_hoistee()
+		return
+
+/obj/structure/hoist/proc/release_hoistee()
+	if(ismob(hoistee))
+		source_hook.unbuckle_mob(hoistee)
+	else
+		hoistee.anchored = FALSE
+	hoistee = null
+	layer = NORMAL_LAYER
```
Comment (comma) [inline code/modules/multiz/hoist.dm]: can just do initial(layer)

**Review by comma — APPROVED** [review]

**Pull request merged**